Fastmodels support: add FVP_MPS2 targets to mbed os

jamesbeyond opened the pull request:
### Description

* add TARGET_ARM_FM top level folder in targets 
* add TARGET_FVP_MPS2 platform category
* add following targets to the FVP_MPS2 category

   * FVP_MPS2_M0
   * FVP_MPS2_M0P
   * FVP_MPS2_M3
   * FVP_MPS2_M4
   * FVP_MPS2_M7

* add all necessary drivers for FVP_MPS2 platfrom and each target
* add toolchian scripts for ARM, GCC_ARM and IAR
* modify target.json enable mbed for fastmodel FVP_MPS2 targets


### Pull request type

    [ ] Fix
    [ ] Refactor
    [x] New target
    [ ] Feature
    [ ] Breaking change

jamesbeyond commented:
@bulislaw Could you review this PR, or add the people who can review this RP. Seems I don't have permission to raise review request

bulislaw commented:
@ARMmbed/mbed-os-maintainers can we add @jamesbeyond to the ArmMbed org please?

bulislaw commented:
@jamesbeyond some checks are failing. Also where did you get all this code from? What about tools changes are they coming in separate PR?

jamesbeyond commented:
@bulislaw Travis failing seem because of Travis server issue. I am not sure how I can re-trigger Travis,
 
This PR only about add the Fastmodel targets to mbed os. 

Most of the code I got from the hardware ARM_MPS2_Mx target drivers, I had some modification to suit the models where necessary 

The changes for the tools will come in separate PRs

adbridge commented:
@bulislaw Qinghao is already a member of mbed-staff..Only a select few people can add reviewers :)

bulislaw reviewed (dismissed):
I've skimmed the files, it looks good.

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,60 @@
+    "FVP_MPS2_M0": {
+        "inherits": ["ARM_FM_Target"],
+        "core": "Cortex-M0",
+        "supported_toolchains": ["GCC_ARM", "ARM", "IAR"],
```
Move to `ARM_FM_Target`

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,60 @@
         "release_versions": ["5"],
         "bootloader_supported": true
+    },
+    "ARM_FM_Target": {
```
Please remove the redundant `Target`

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,60 @@
+        "inherits": ["ARM_FM_Target"],
+        "core": "Cortex-M0",
+        "supported_toolchains": ["GCC_ARM", "ARM", "IAR"],
+        "extra_labels": ["ARM_FM", "FVP_MPS2", "FVP_MPS2_M0"],
```
Move "ARM_FM" (should be the parent name) and `"FVP_MPS2"` to the parent, and change this to an `extra_labels_add`

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,60 @@
+        "core": "Cortex-M0",
+        "supported_toolchains": ["GCC_ARM", "ARM", "IAR"],
+        "extra_labels": ["ARM_FM", "FVP_MPS2", "FVP_MPS2_M0"],
+        "OUTPUT_EXT": "elf",
```
Move to parent.

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,60 @@
+        "extra_labels": ["ARM_FM", "FVP_MPS2", "FVP_MPS2_M0"],
+        "OUTPUT_EXT": "elf",
+        "macros": ["CMSDK_CM0","CMSIS_VECTAB_VIRTUAL", "CMSIS_VECTAB_VIRTUAL_HEADER_FILE=\"cmsis_nvic.h\""],
+        "device_has": ["AACI", "ANALOGIN", "CLCD", "I2C", "INTERRUPTIN", "PORTIN", "PORTINOUT", "PORTOUT", "SERIAL", "SERIAL_FC", "SPI", "SPISLAVE", "TSC"],
```
Don't duplicate the devices_has of the parent. If anything is added here, add it with `extra_labels_add`.

Commit:
FastModels: add a new parent platform ARM_FM

create a new platform folder TARGET_ARM_FM
add general drivers files for FVP_MPS2

Commit:
FastModels: add FVP_MPS2_M3 target support

add cmsis drivers and toolchain scripts for FVP_MPS2_M3

Commit:
FastModels: add FVP_MPS2_M4 target support

add cmsis drivers and toolchain scripts for FVP_MPS2_M4

Commit:
FastModels: add FVP_MPS2_M0 target support

add cmsis drivers and toolchain scripts for FVP_MPS2_M0

Commit:
FastModels: add FVP_MPS2_M0P target support

add cmsis drivers and toolchain scripts for FVP_MPS2_M0P

Commit:
FastModels: add FVP_MPS2_M7 target support

add cmsis drivers and toolchain scripts for FVP_MPS2_M7

Commit:
FastModels: enable FVP_MPS2 targets for mbed os

add ARM_FM top level category
add second level parent FVP_MPS2 inherits ARM_FM
add each target ( M0 M0P M3 M4 M7 ) inherits FVP_MPS2

jamesbeyond commented:
Hi @theotherjimmy, all of your comments been addressed by adding a new parent FVP_MPS2. please review again

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/PeripheralNames.h:
```diff
@@ -0,0 +1,91 @@
+} ADCName;
+
+typedef enum {
+    SPI_0 = (int)MPS2_SSP1_BASE,
```
Inconsistent indentation levels

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/PinNames.h:
```diff
@@ -0,0 +1,243 @@
+#endif
+
+typedef enum {
+        PIN_INPUT,
```
Should be 4 spaces

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/PinNames.h:
```diff
@@ -0,0 +1,243 @@
+
+typedef enum {
+    // MPS2 EXP Pin Names
+        EXP0 = 0,
```
should be 4 spaces

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+
+    error = 0;
+    val = SMSC9220->MAC_CSR_CMD;
+    if(!(val & ((unsigned int)1 << 31))) {    // Make sure there's no pending operation
```
Check the guidelines to see if there is supposed to be a space after 'if'

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+        if(!timedout) {
+            error = 1;
+        }
+        else
```
should be  - 

if(!timedout) {
    error = 1;
} else {
   ...
}

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+            error = 1;
+        }
+    } else {
+       printf("Warning: SMSC9220 MAC CSR is busy. No data written.\n");
```
Why is there a printf here ?

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+        if(!timedout) {
+            error = 1;
+        }
+        else
```
Should be } else {

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+        }
+
+    } else {
+        printf("Warning: SMSC9220 MAC MII is busy. No data written.\n");
```
Printf ?

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+    } while(timedout && (SMSC9220->HW_CFG & 1));
+
+    if(!timedout)
+        return 1;
```
Check guidelines about whether { } are always required even for single statements or not

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+unsigned int smsc9220_reset_phy(void)
+{
+    unsigned short read;
+    int error;
```
This variable is redundant

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+
+    smsc9220_phy_regwrite(SMSC9220_PHY_ANEG_ADV, aneg_adv);
+    smsc9220_phy_regread(SMSC9220_PHY_ANEG_ADV, &aneg_adv);
+    return;
```
why return without a value ?

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+        SMSC9220->HW_CFG = hw_cfg;
+    }
+
+    return;
```
Why return without a value ? (check all instances of this)

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -0,0 +1,450 @@
+    xmit_inf = SMSC9220->TX_FIFO_INF;
+
+    if(xmit_stat2 != 0 ) {
+        for(i = 0; i < 6; i++) {
```
what does the '6' mean ?

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/fpga.c:
```diff
@@ -0,0 +1,90 @@
+    unsigned int start;
+
+    start = MPS2_FPGAIO->COUNTER;
+    end   = start + (tick);
```
spacing wrong

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/fpga.c:
```diff
@@ -0,0 +1,90 @@
+    {
+        while (MPS2_FPGAIO->COUNTER >= start && MPS2_FPGAIO->COUNTER < end);
+    }
+    else
```
} else {

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/fpga.c:
```diff
@@ -0,0 +1,90 @@
+    {
+        while (MPS2_FPGAIO->COUNTER >= start && MPS2_FPGAIO->COUNTER < end);
+    }
+    else
```
} else {

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/fpga.c:
```diff
@@ -0,0 +1,90 @@
+    }
+    else
+    {
+        while (MPS2_FPGAIO->COUNTER >= start);
```
I assume these are volatile registers ?

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/fpga.c:
```diff
@@ -0,0 +1,90 @@
+    unsigned int start;
+
+    start = MPS2_FPGAIO->COUNTER;
+    end   = start + (25 * usec);
```
spacing wrong

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/fpga.c:
```diff
@@ -0,0 +1,90 @@
+    {
+        while (MPS2_FPGAIO->COUNTER >= start && MPS2_FPGAIO->COUNTER < end);
+    }
+    else
```
} else {

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/mps2_ethernet_api.c:
```diff
@@ -0,0 +1,166 @@
+
+    if(mac_high != mac_valid_high || mac_low != mac_valid_low) {
+        error = TRUE;
+        return error;
```
This line is redundant

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/analogin_api.c:
```diff
@@ -0,0 +1,110 @@
+void analogin_init(analogin_t *obj, PinName pin) {
+    
+    obj->adc = (ADCName)pinmap_peripheral(pin, PinMap_ADC);
+  MBED_ASSERT(obj->adc != (ADCName)NC);
```
indentation wrong

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/analogin_api.c:
```diff
@@ -0,0 +1,110 @@
+    obj->address = (0x0000 | (pin-600));
+    
+    SPIName adc_mosi = (SPIName)pinmap_peripheral(ADC_MOSI, PinMap_SPI_MOSI);
+  SPIName adc_miso = (SPIName)pinmap_peripheral(ADC_MISO, PinMap_SPI_MISO);
```
indentation

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/analogin_api.c:
```diff
@@ -0,0 +1,110 @@
+  obj->adc_spi = (MPS2_SSP_TypeDef*)pinmap_merge(adc_data, adc_cntl);
+  
+    if(analog_spi_inited == 0){
+    obj->adc_spi->CR1       = 0;
```
indentation is horrible

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/device.h:
```diff
@@ -0,0 +1,43 @@
+
+
+
+
```
why all the blank lines ?

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/ethernet_api.c:
```diff
@@ -0,0 +1,164 @@
+ *----------------------------------------------------------------------------*/
+int ethernet_init()
+{
+      int error;
```
indentation

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/ethernet_api.c:
```diff
@@ -0,0 +1,164 @@
+
+    if(smsc9220_reset_phy()) {
+        error = TRUE;
+        return error;
```
why does this error return immediately but the ones above don't? Also if returning immediately why is the error variable even used ? Why not just return TRUE ?

jamesbeyond replied:
Through this code was not originally from me, I guess that is for readability.
let people know this returns an error value, rather than miss leading people by TRUE implicates successful.

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/ethernet_api.c:
```diff
@@ -0,0 +1,164 @@
+
+int ethernet_write(const char *data, int size)
+{
+        return 0;
```
indentation

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/ethernet_api.c:
```diff
@@ -0,0 +1,164 @@
+
+int ethernet_send() 
+{
+ return 0;
```
indentation

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/gpio_api.c:
```diff
@@ -0,0 +1,140 @@
+/* mbed Microcontroller Library
```
Formatting in this file is horrible, please conform to the coding guidelines

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/gpio_irq_api.c:
```diff
@@ -0,0 +1,241 @@
+/* mbed Microcontroller Library
```
File needs reformatting to meet coding standards

adbridge reviewed (changes requested):
The level of formatting in the majority of these files does not meet the coding guidelines, they all need to be tidied up.

adbridge commented:
@theotherjimmy can you re-review the changes to the target stuff on Monday when you are back please

Commit:
reformat coding styles based on coding guidelines

jamesbeyond commented:
Hi @adbridge, all of your comments about formatting been addressed, Could you review again?

adbridge commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/SDK/ETH_MPS2.c:
```diff
@@ -61,10 +61,9 @@ unsigned int smsc9220_mac_regread(unsigned char regoffset, unsigned int *data)
             error = 1;
-        }
-        else
+        } else
```
{ } needed around single statements

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,49 @@
+        "inherits": ["ARM_FM"],
+        "public": false,
+        "supported_toolchains": ["GCC_ARM", "ARM", "IAR"],
+        "extra_labels_add": ["FVP_MPS2"],
```
Redundant. This already happens because of the target name.

jamesbeyond replied:
Didn't realize that was automatically added. Now got all of these redundant "extra_label_add" removed. can you review pls @theotherjimmy

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,49 @@
+    "FVP_MPS2_M0": {
+        "inherits": ["FVP_MPS2"],
+        "core": "Cortex-M0",
+        "extra_labels_add": ["FVP_MPS2_M0"],
```
Redundant. This already happens because of the target name.

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,49 @@
+    "FVP_MPS2_M0P": {
+        "inherits": ["FVP_MPS2"],
+        "core": "Cortex-M0+",
+        "extra_labels_add": ["FVP_MPS2_M0P"],
```
Redundant. This already happens because of the target name.

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,49 @@
+    "FVP_MPS2_M3": {
+        "inherits": ["FVP_MPS2"],
+        "core": "Cortex-M3",
+        "extra_labels_add": ["FVP_MPS2_M3"],
```
Redundant. This already happens because of the target name.

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,49 @@
+    "FVP_MPS2_M4": {
+        "inherits": ["FVP_MPS2"],
+        "core": "Cortex-M4",
+        "extra_labels_add": ["FVP_MPS2_M4"],
```
Redundant. This already happens because of the target name.

theotherjimmy commented on targets/targets.json:
```diff
@@ -4008,5 +4008,49 @@
+    "FVP_MPS2_M7": {
+        "inherits": ["FVP_MPS2"],
+        "core": "Cortex-M7",
+        "extra_labels_add": ["FVP_MPS2_M7"],
```
Redundant. This already happens because of the target name.

Commit:
update targets.json remove redundant extra_lables

Commit:
remove empty lines and redundant variables

theotherjimmy reviewed (dismissed):
Looks good.

Commit:
remove ethernet driver as it is not compatible

jamesbeyond commented:
@adbridge could you review again pls

adbridge commented:
/morph build

0xc0170 commented on targets/TARGET_ARM_FM/TARGET_FVP_MPS2/serial_api.c:
```diff
@@ -0,0 +1,384 @@
+{
+}
+void serial_set_flow_control(serial_t *obj, FlowControl type, PinName rxflow, PinName txflow)
+{
```
why flow control is emtpy? is it enabled? will it be supported?

Similar is to pinmap functionality (assume that target does not have any pinmap functionality, neither break set/clear) thus they are emptY?

jamesbeyond replied:
the fastmodels and the hardware it modeling after don't support for flow control.
The details can be found in bellow reference guide:
*  Ref. [DDI0479C_cortex_m_system_design_kit_r1p0_trm.pdf ](http://infocenter.arm.com/help/topic/com.arm.doc.ddi0479c/DDI0479C_cortex_m_system_design_kit_r1p0_trm.pdf)

0xc0170 reviewed (approved):
Just one question regarding empty functions in HAL.

0xc0170 commented:
@studavekar Please review the latest build (`[Errno 28] No space left on device` error)

jamesbeyond commented:
Hi @studavekar, any update on this build error?

cmonr commented:
Should be resolved now.
/morph build
